File: gin/library/redis.go
package library

import (
	"fmt"
	"time"
	"webmis/config"

	redigo "github.com/gomodule/redigo/redis"
)

var RedisDB *redigo.Pool      //默认池
var RedisDBOther *redigo.Pool //其它池

/* 缓存数据库 */
type Redis struct {
	db   string      //数据库
	conn redigo.Conn //连接
}

/* 构造函数 */
func (r *Redis) New(db string) *Redis {
	r.db = db
	r.RedisConn()
	return r
}

/* 连接 */
func (r *Redis) RedisConn() *redigo.Conn {
	if r.db == "other" {
		if RedisDBOther == nil {
			RedisDBOther = r.RedisPool(config.RedisOther())
		}
		r.conn = RedisDBOther.Get()
	} else {
		if RedisDB == nil {
			RedisDB = r.RedisPool(config.Redis())
		}
		r.conn = RedisDB.Get()
	}
	return &r.conn
}

/* 关闭 */
func (r *Redis) Close() {
	if r.conn != nil {
		r.conn.Close()
	}
}

/* 数据池 */
func (r *Redis) RedisPool(cfg *config.RedisType) *redigo.Pool {
	return &redigo.Pool{
		MaxIdle:         cfg.Min,  //空闲数
		MaxActive:       cfg.Max,  //最大数
		IdleTimeout:     cfg.Time, //空闲超时(秒)
		MaxConnLifetime: cfg.Time, //最大连接时长(秒)
		Wait:            true,     //超过最大连接数: true 等待 false 报错
		// 连接
		Dial: func() (redigo.Conn, error) {
			conn, err := redigo.Dial(cfg.Driver, cfg.Host+":"+cfg.Port)
			if err != nil {
				fmt.Println("[Redis] Conn:", err)
				return nil, err
			}
			// 密码
			if cfg.Password != "" {
				conn.Do("AUTH", cfg.Password)
			}
			// 硬盘
			conn.Do("SELECT", cfg.Db)
			return conn, err
		},
		// 检测
		TestOnBorrow: func(c redigo.Conn, t time.Time) error {
			if time.Since(t) < time.Minute {
				return nil
			}
			_, err := c.Do("PING")
			return err
		},
	}
}

/* 添加 */
func (r *Redis) Set(val ...interface{}) []byte {
	res, err := redigo.Bytes(r.conn.Do("SET", val...))
	if err != nil {
		fmt.Println("[Redis] Set:", err)
		return nil
	}
	return res
}

/* 自增 */
func (r *Redis) Incr(key string) int64 {
	res, err := redigo.Int64(r.conn.Do("Incr", key))
	if err != nil {
		fmt.Println("[Redis] Incr:", err)
		return 0
	}
	return res
}

/* 获取 */
func (r *Redis) Get(key string) string {
	res, err := redigo.String(r.conn.Do("Get", key))
	if err != nil {
		fmt.Println("[Redis] Get:", err)
		return ""
	}
	return res
}

/* 删除 */
func (r *Redis) Del(keys ...interface{}) bool {
	res, err := redigo.Bool(r.conn.Do("DEL", keys...))
	if err != nil {
		fmt.Println("[Redis] Del:", err)
		return false
	}
	return res
}

/* 是否存在 */
func (r *Redis) Exist(key string) bool {
	res, err := redigo.Bool(r.conn.Do("EXISTS", key))
	if err != nil {
		fmt.Println("[Redis] Exist:", err)
		return false
	}
	return res
}

/* 设置过期时间(秒) */
func (r *Redis) Expire(key string, ttl int64) int64 {
	res, err := redigo.Int64(r.conn.Do("EXPIRE", key, ttl))
	if err != nil {
		fmt.Println("[Redis] Expire:", err)
		return 0
	}
	return res
}

/* 获取过期时间(秒) */
func (r *Redis) TTL(key string) int64 {
	res, err := redigo.Int64(r.conn.Do("TTL", key))
	if err != nil {
		fmt.Println("[Redis] TTL:", err)
		return 0
	}
	return res
}

/* 获取长度 */
func (r *Redis) StrLen(key string) int64 {
	res, err := redigo.Int64(r.conn.Do("STRLEN", key))
	if err != nil {
		fmt.Println("[Redis] StrLen:", err)
		return 0
	}
	return res
}

/* 哈希(Hash)-添加 */
func (r *Redis) HSet(name string, key string, val interface{}) int64 {
	res, err := redigo.Int64(r.conn.Do("HSET", name, key, val))
	if err != nil {
		fmt.Println("[Redis] HSet:", err)
		return 0
	}
	return res
}
func (r *Redis) HMSet(name string, obj interface{}) int64 {
	res, err := redigo.Int64(r.conn.Do("HMSET", redigo.Args{}.Add(name).AddFlat(&obj)...))
	if err != nil {
		fmt.Println("[Redis] HMSet:", err)
		return 0
	}
	return res
}

/* 哈希(Hash)-获取 */
func (r *Redis) HGet(name string, key string) []byte {
	res, err := redigo.Bytes(r.conn.Do("HGET", name, key))
	if err != nil {
		fmt.Println("[Redis] HGet:", err)
		return nil
	}
	return res
}
func (r *Redis) HMGet(name string, keys ...string) []interface{} {
	args := []interface{}{name}
	for _, field := range keys {
		args = append(args, field)
	}
	res, err := redigo.Values(r.conn.Do("HMGET", args))
	if err != nil {
		fmt.Println("[Redis] HMGet:", err)
		return nil
	}
	return res
}

/* 哈希(Hash)-删除 */
func (r *Redis) HDel(name string, key ...string) int64 {
	res, err := redigo.Int64(r.conn.Do("HDEL", name, key))
	if err != nil {
		fmt.Println("[Redis] HDel:", err)
		return 0
	}
	return res
}

/* 哈希(Hash)-是否存在 */
func (r *Redis) HExist(name string, key string) bool {
	res, err := redigo.Bool(r.conn.Do("HEXISTS", name, key))
	if err != nil {
		fmt.Println("[Redis] HExist:", err)
		return false
	}
	return res
}

/* 哈希(Hash)-Key个数 */
func (r *Redis) HLen(name string) int64 {
	res, err := redigo.Int64(r.conn.Do("HLEN", name))
	if err != nil {
		fmt.Println("[Redis] HLen:", err)
		return 0
	}
	return res
}

/* 列表(List)-写入 */
func (r *Redis) RPush(key string, val interface{}) bool {
	res, err := redigo.Bool(r.conn.Do("RPUSH", key, val))
	if err != nil {
		fmt.Println("[Redis] LPUSH:", err)
		return false
	}
	return res
}
func (r *Redis) LPush(key string, val interface{}) bool {
	res, err := redigo.Bool(r.conn.Do("LPUSH", key, val))
	if err != nil {
		fmt.Println("[Redis] LPUSH:", err)
		return false
	}
	return res
}

/* 列表(List)-读取 */
func (r *Redis) LRange(key string, start int64, end int64) []interface{} {
	res, err := redigo.Values(r.conn.Do("LRange", key, start, end))
	if err != nil {
		return nil
	}
	return res
}
func (r *Redis) BRPop(key string, timeout int) []interface{} {
	res, err := redigo.Values(r.conn.Do("BRPOP", key, timeout))
	if err != nil {
		return nil
	}
	return res
}
func (r *Redis) BLPop(key string, timeout int) []interface{} {
	res, err := redigo.Values(r.conn.Do("BLPOP", key, timeout))
	if err != nil {
		return nil
	}
	return res
}

File: java/src/main/java/webmis/library/Redis.java
package webmis.library;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.JedisPoolConfig;
import webmis.service.Base;

/* 缓存数据库 */
public class Redis extends Base {

  public static JedisPool RedisDB = null;       //默认池
  public static JedisPool RedisDBOther = null;  //其它池

  private String _db = "";                      //数据库
  private Jedis _conn = null;                   //连接

  /* 构造函数 */
  public Redis(String db) {
    _db = db;
    RedisConn();
  }

  /* 连接 */
  public Jedis RedisConn() {
    HashMap<String, Object> cfg;
    try {
      if(_db.equals("other")) {
        cfg=webmis.config.Redis.Other();
        if(Redis.RedisDBOther==null) Redis.RedisDBOther=RedisPool(cfg);
        _conn = Redis.RedisDBOther.getResource();
      } else {
        cfg=webmis.config.Redis.Default();
        if(Redis.RedisDB==null) Redis.RedisDB=RedisPool(cfg);
        _conn = Redis.RedisDB.getResource();
      }
      // 密码
      String passwd = cfg.get("password").toString();
      if(!passwd.isEmpty()) _conn.auth(passwd);
      // 硬盘
      _conn.select(Integer.valueOf(cfg.get("db").toString()));
      return _conn;
    } catch (Exception e) {
      Print("[Redis] Conn:", e.getMessage());
      return null;
    }
  }

  /* 关闭 */
  public void Close() {
    if(_conn!=null && _conn.isConnected()) _conn.close();
  }

  /* 连接池 */
  public JedisPool RedisPool(HashMap<String, Object> cfg) {
    try{
      JedisPoolConfig config = new JedisPoolConfig();
      config.setMaxIdle(Integer.valueOf(cfg.get("min").toString()));
      config.setMaxTotal(Integer.valueOf(cfg.get("max").toString()));
      JedisPool pool = new JedisPool(config, cfg.get("host").toString(), Integer.valueOf(cfg.get("port").toString()));
      return pool;
    } catch (Exception e) {
      Print("[Model] Pool:", e.getMessage());
      return null;
    }
  }

  /* 添加 */
  public String Set(String key, String val) {
    if(_conn==null) return "";
    return _conn.set(key, val);
  }
  /* 自增 */
  public Long Incr(String key) {
    if(_conn==null) return null;
    return _conn.incr(key);
  }
  /* 获取 */
  public String Get(String key) {
    if(_conn==null) return "";
    return _conn.get(key);
  }
  /* 删除 */
  public long Del(String... key) {
    if(_conn==null) return 0;
    return _conn.del(key);
  }
  /* 是否存在 */
  public boolean Exist(String key) {
    if(_conn==null) return false;
    return _conn.exists(key);
  }
  /* 设置过期时间(秒) */
  public long Expire(String key, int ttl) {
    if(_conn==null) return 0;
    return _conn.expire(key, ttl);
  }
  /* 获取过期时间(秒) */
  public long Ttl(String key) {
    if(_conn==null) return 0;
    return _conn.ttl(key);
  }
  /* 获取长度 */
  public long StrLen(String key) {
    if(_conn==null) return 0;
    return _conn.strlen(key);
  }

  /* 哈希(Hash)-添加 */
  public Long HSet(String name, String key, String val) {
    if(_conn==null) return null;
    return _conn.hset(name, key, val);
  }
  public String HMSet(String name, Map<String, String> obj) {
    if(_conn==null) return "";
    return _conn.hmset(name, obj);
  }
  /* 哈希(Hash)-获取 */
  public String HGet(String name, String key) {
    if(_conn==null) return "";
    return _conn.hget(name, key);
  }
  public List<String> HMGet(String name, String... key) {
    if(_conn==null) return null;
    return _conn.hmget(name, key);
  }
  /* 哈希(Hash)-删除 */
  public long HDel(String name, String... key) {
    if(_conn==null) return 0;
    return _conn.hdel(name, key);
  }
  /* 哈希(Hash)-是否存在 */
  public boolean HExist(String name, String key) {
    if(_conn==null) return false;
    return _conn.hexists(name, key);
  }
  /* 哈希(Hash)-Key个数 */
  public long HLen(String name) {
    if(_conn==null) return 0;
    return _conn.hlen(name);
  }

  /* 列表(List)-写入 */
  public long RPush(String key, String... val) {
    if(_conn==null) return 0;
    return _conn.rpush(key, val);
  }
  public long LPush(String key, String... val) {
    if(_conn==null) return 0;
    return _conn.lpush(key, val);
  }

  /* 列表(List)-读取 */
  public List<String> LRange(String key, int start, int end) {
    if(_conn==null) return null;
    return _conn.lrange(key, start, end);
  }
  public List<String> BRPop(String key, int timeout) {
    if(_conn==null) return null;
    return _conn.brpop(timeout, key);
  }
  public List<String> BLPop(String key, int timeout) {
    if(_conn==null) return null;
    return _conn.blpop(timeout, key);
  }

}
